openghg/retrieve/_search_helpers.py

```diff
@@ -0,0 +1,138 @@
+
+
+def _in_interval(x: T, start: Optional[T], stop: Optional[T]) -> bool:
+    """Return True if start <= x <= stop; if either start or stop is None, omit
+    the corresponding bound.
+
+    Args:
+        x: value to test for inclusion in the interval [start, stop]
+        start: start of the interval
+        stop: end of the interval
+
+    Returns:
+        True if start <= x <= stop, False otherwise
+    """
+    if start is None and stop is None:
+        return False
+    elif start is None:
+        stop = cast(T, stop)  # to appease mypy
+        return x <= stop
+    elif stop is None:
+        return x >= start
+    else:
+        return start <= x <= stop
+
+
+def _convert_slice_to_test(s: slice, key: Optional[str] = None) -> Callable:
+    """Convert slice to a function that checks if values are in the interval specified by the slice.
+
+    Args:
+        s: slice specifying interval of values
+        key: optional metadata key, used to choose value formatting
+
+    Returns:
+        function that returns True if values are in the slice interval
+    """
+
+    def formatter(x: Any) -> Any:
+        """Formatting if key == 'inlet': try to extract float from strings."""
+        if key == "inlet":
+            if isinstance(x, (int, float)):
+                return x
+            if isinstance(x, str):
+                try:
+                    result = extract_float(x)
+                except ValueError:
+                    return None
+                else:
+                    return result
+            return None
+        return x  # key != "inlet"
+
+    def test_func(x: Any) -> bool:
+        """Return True if start <= formatter(x) <= stop."""
+        return _in_interval(formatter(x), formatter(s.start), formatter(s.stop))  # type: ignore
+
+    return test_func
+
+
+def process_special_queries(search_terms: dict, neg_lookup_flag: Any = "NOT_SET_FLAG") -> dict:
+    """Separate test queries and negative lookup keys from normal search terms.
+
+    Args:
+        search_terms: dict of search terms
+        neg_lookup_flag: flag value used to indicate negative lookup on a key
+
+    Returns:
+        dict containing search_terms dict, search_tests dict, and negative_lookup_keys list, which
+        are the parameters for TinyDBMetastore.search
+    """
+    _search_terms = search_terms.copy()  # copy to avoid mutating search_terms while iterating over items
+    search_tests = {}
+    negative_lookup_keys = []
+
+    for k, v in search_terms.items():
+        if isinstance(v, slice):
+            search_tests[k] = _convert_slice_to_test(v, key=k)
+            del _search_terms[k]
+        elif v == neg_lookup_flag:
+            negative_lookup_keys.append(k)
+            del _search_terms[k]
+
+    return {
+        "search_terms": _search_terms,
+        "search_tests": search_tests,
+        "negative_lookup_keys": negative_lookup_keys,
+    }
+
+
+def parse_search_kwargs(search_kwargs: dict) -> list[dict]:
```

Comment:
A small thought - should we reserve the term `parse` in our codebase for the standardise parsers? In which case could this be `format_search_terms` (or something else)?

Reply:
I changed it to `flatten_search_terms`. Maybe `process_search_terms` is better? I guess flattening and formatting both happen...